admin_panel/bd_models/admin/player.py

```diff
@@ -101,3 +101,39 @@ def blacklist_users(self, request: "HttpRequest", queryset: "QuerySet[Player]",
             f"{request.user} blacklisted players "
             f'{", ".join([str(x.discord_id) for x in queryset])} for the reason: {data["reason"]}.'
         )
+
+    @action_with_form(
+        UnblacklistActionForm, description="Unblacklist the selected users"
+    )  # type: ignore
+    def unblacklist_users(self, request: "HttpRequest", queryset: "QuerySet[Player]", data: dict):
+        reason = (
+            data["reason"]
+            + f"\nDone through the admin panel by {request.user} ({request.user.pk})"
+        )
+        blacklists: list[BlacklistedID] = []
+        histories: list[BlacklistHistory] = []
+        for player in queryset:
+            if not BlacklistedID.objects.filter(discord_id=player.discord_id).exists():
+                self.message_user(
+                    request, f"Player {player.discord_id} is not blacklisted!", messages.ERROR
+                )
+                return
+            blacklists.append(
+                BlacklistedID(discord_id=player.discord_id, reason=reason, moderator_id=None)
+            )
+            histories.append(
+                BlacklistHistory(discord_id=player.discord_id, reason=reason, moderator_id=0)
+            )
+        BlacklistedID.objects.filter(discord_id__in=[b.discord_id for b in blacklists]).delete()
+        BlacklistHistory.objects.bulk_create(histories)
```

Comment:
This code doesn't work since you're filtering the blacklists by reason, so unless your unblacklist reason is the exact same as the blacklist reason, it's not going to match anything. This code only saves the IDs instead. You also missed setting the `unblacklist` type in history.
```suggestion
        blacklists: list[int] = []
        histories: list[BlacklistHistory] = []
        for player in queryset:
            blacklist = BlacklistedID.objects.filter(discord_id=player.discord_id).first()
            if not blacklist:
                self.message_user(
                    request, f"Player {player.discord_id} is not blacklisted!", messages.ERROR
                )
                return
            blacklists.append(blacklist.id)
            histories.append(
                BlacklistHistory(
                    discord_id=player.discord_id,
                    reason=reason,
                    moderator_id=0,
                    action_type="unblacklist",
                )
            )
        BlacklistedID.objects.filter(id__in=blacklists).delete()
        BlacklistHistory.objects.bulk_create(histories)
```